Refactor pospell to use multiprocessing

rtobar opened the pull request:
One of the main drawbacks of pospell at the moment is that checking is
performed serially by a single hunspell process. In small projects this
is not noticeable, but in slightly bigger ones this can go up a bit
(e.g., in python-docs-es it takes ~2 minutes to check the whole set of
.po files).

The obvious solution to speed things up is to use multiprocessing,
parallelising the process at two different places: first, when reading
the input .po files and collecting the input strings to feed into
hunspell, and secondly when running hunspell itself.

This commit implements this support. It works as follows:

 * A new namedtuple called input_line has been added. It contains a
   filename, a line, and text, and thus it uniquely identifies an input
   line in a self-contained way.
 * When collecting input to feed into hunspell, the po_to_text routine
   collects input_lines instead of a simple string. This is done with a
   multiprocessing Pool to run in parallel across all input files.
 * The input_lines are split in N blocks, with N being the size of the
   pool. Note that during this process input_lines from different files
   might end up in the same block, and input_lines from the same file
   might end up in different blocks; however since input_lines are
   self-contained we are not losing information.
 * N hunspell instances are run over the N blocks of input_lines using
   the pool (only the text field from the input_lines is fed into
   hunspell).
 * When interpreting errors from hunspell we can match an input_line
   with its corresponding hunspell output lines, and thus can identify
   the original file:line that caused the error.

The multiprocessing pool is sized via a new -j/--jobs command line
option, which defaults to os.cpu_count() to run at maximum speed by
default.

These are the kind of differences I see with python-docs-es in my
machine, so YMMV depending on your setup/project:

```
$> time pospell -p dict2.txt -l es_ES */*.po -j 1
real    2m1.859s
user    2m6.680s
sys     0m3.829s

$> time pospell -p dict2.txt -l es_ES */*.po -j 2
real    1m10.322s
user    2m18.210s
sys     0m3.559s
```

Signed-off-by: Rodrigo Tobar <rtobar@icrar.org>

JulienPalard commented on pospell.py:
```diff
@@ -202,18 +209,18 @@ def quote_for_hunspell(text):
 
 
-def po_to_text(po_path, drop_capitalized=False):
+def po_to_text(drop_capitalized, po_path):
```
Changing the order of the arguments of this function was not mandatory, you could have used a keyword argument to `partial`. But I think it's safe to bet that nobody used the function out of pospell, so I'm OK if you leave it as is too.

rtobar replied:
I don´t use `partial` that often, so I didn't know/remember you could pass down keyword arguments down, sweet! I'll keep the old signature -- although this is a mid-size refactoring I really prefer it to be as less invasive as possible.

JulienPalard commented on pospell.py:
```diff
@@ -329,67 +365,51 @@ def spell_check(
-            print(po_to_text(str(po_file), drop_capitalized))
-            continue
-        texts_for_hunspell[po_file] = po_to_text(str(po_file), drop_capitalized)
+    pool = multiprocessing.Pool(jobs)
```
Please use a context manager here, else I'll have:
```
/usr/lib/python3.9/multiprocessing/pool.py:265: ResourceWarning: unclosed running multiprocessing pool <multiprocessing.pool.Pool state=RUN pool_size=16>
  _warn(f"unclosed running multiprocessing pool {self!r}",
ResourceWarning: Enable tracemalloc to get the object allocation traceback
```

rtobar replied:
will do!

JulienPalard reviewed (changes requested):
🚀🚀🚀 On my laptop it goes from 1m54s to 14s. 🚀🚀🚀

A few issues though, some I commented (the order of parameters is nitpicking, the ResourceWarning I'm really not a fan), some you'll find by running `tox -p all` on your laptop. If you love multiprocessing, you'll love the `-p all` of tox ;)

About the number of parameter, just add `--max-args=6` to the pylint command in `tox.ini`, don't refactor everything just to save one argument, I don't care.

Please give something else than `None` by default to `jobs` in `spell_check`: it fails with "TypeError: unsupported operand types(s) for +: `int` and `NoneType`" when called without jobs, with I do in the tests (try running `python -m pytest`). I'm OK with `jobs=os.cpu_count()` as a default argument.

After fixing the jobs parameter for the tests, I still have a few tests failing, could you take a look?

JulienPalard commented:
The two-step thing is really interesting, it balances well the load between CPUs, nice job. My first jugdment was "Oh but at the end of step 1 some CPUs are idle, it may be even faster to do it in a single pass.". I even tried it out of curiosity:

```python
def spell_check_one(po_file, personal_dict=None, language="en_US", drop_capitalized=False, debug_only=False,):
    input_lines = po_to_text(drop_capitalized, po_file)
    if debug_only:
        for filename, line, text in input_lines:
            print(filename, line, text, sep=":")
        return 0
    return run_hunspell(language, personal_dict, input_lines)


def spell_check(po_files, personal_dict=None, language="en_US", drop_capitalized=False, debug_only=False, jobs=1,):
    with multiprocessing.Pool(jobs) as pool:
        errors = flatten(
            pool.map(
                functools.partial(
                    spell_check_one,
                    personal_dict=personal_dict,
                    language=language,
                    drop_capitalized=drop_capitalized,
                    debug_only=debug_only,
                ),
                po_files,
            )
        )
    for error in errors:
        print(*error, sep=":")
    return len(errors)
```

But mine runs in 19s while yours run in 15s, you win ;)

rtobar commented:
> rocketrocketrocket On my laptop it goes from 1m54s to 14s. rocketrocketrocket

Awesome! :)

> A few issues though, some I commented (the order of parameters is nitpicking, the ResourceWarning I'm really not a fan), some you'll find by running `tox -p all` on your side. If you love multiprocessing, you'll love the `-p all` of tox ;)

Grrrr, rookie mistake from my part not to check the tests! I guess I focused too much on manually testing against our project that I forgot to check if there was a testing infrastructure set up in the pospell itself. I'll go through all the points now, specially those you highlighted in your inline comments.

> About the number of parameter, just add `--max-args=6` to the pylint command in `tox.ini`, don't refactor everything just to save one argument, I don't care.

Will do!

> Please give something else than `None` by default to `jobs` in `spell_check`: it fails with "TypeError: unsupported operand types(s) for +: `int` and `NoneType`" when called without jobs, with I do in the tests (try running `python -m pytest`, or `tox -e py39`). I'm OK with `jobs=os.cpu_count()` as a default argument here.

Again, that's me not running the tests, I'll give a more sensible default value

rtobar commented:
I've just force-pushed a new version with the following updates:

 * Ran black over the code
 * The pool's lifecycle is now properly handled. I had to manually call `close()` as `__exit__` calls `terminate()`, which doesn't allow subprocess coverage collection work reliably.
 * Put back `po_to_text`'s signature to the way it was
 * Addressed all pylint/mypy/flake8/pydocstyle warnings
 * Defaulted `check_spell`'s `job` parameter to `os.cpu_count()`
 * Fixed `check_spell` for when there is no input to translate (avoiding even having to spawn `hunspell`, which wasn't the case previously I believe).

JulienPalard commented:
I just reverted the change to use pytest-cov, which I try to avoid because it's not compatible with `tox -p all` (IIRC it's because pytest-cov runs `coverage combine` during each runs, see https://github.com/pytest-dev/pytest-cov/issues/398). It does not fail often, it's a rare case (when two tested envs of tox does the combine at the same moment) but I don't like randomly failing things.

Also I don't think replacing `coverage run -m pytest` with `pytest-cov` is worth 1k lines of code (`e=mc²` he said, `errors = (more code)²`).

JulienPalard commented:
Thanks @rtobar for this great PR \o/

JulienPalard commented:
This is released on pypi as pospell 1.1.

rtobar commented:
@JulienPalard thanks to you for the very prompt replies, insightful comments, and for the new release. Let's enjoy some speed :+1:
